Fixes #12193 - Stop using counter_cache

tbrisker opened the pull request:
This is aimed to prevent deadlocks caused by Rails' implementation of
counter_cache that is hit when multiple hosts are created or updated at
the same time. An added benefit is that the counters will now display
the number of hosts scoped to the current taxonomy and authorized for
the user to see. It also removes a lot of code used to workaround
various counter_cache issues.
The downsides are that sorting and filtering based on host counts will
no longer work, and that page load times may increase for pages
containing host counters, such as enviroments index.

tbrisker committed:
Fixes #12193 - Stop using counter_cache for host[group] counts

This is aimed to prevent deadlocks caused by Rails' implementation of
counter_cache that is hit when multiple hosts are created or updated at
the same time. An added benefit is that the counters will now display
the number of hosts scoped to the current taxonomy and authorized for
the user to see. It also removes a lot of code used to workaround
various counter_cache issues.
The downsides are that sorting and filtering based on host counts will
no longer work, and that page load times increase slightly for pages
containing host counters, such as enviroments index.

tbrisker commented:
@beav @lzap would be happy if you could take a look. This still needs testing on a very large DB to check how long the various pages take to load.

ohadlevy commented on app/controllers/models_controller.rb:
```diff
@@ -1,6 +1,7 @@
 class ModelsController < ApplicationController
   include Foreman::Controller::AutoCompleteSearch
   before_filter :find_resource, :only => [:edit, :update, :destroy]
+  before_filter :find_hosts_count, :only => :index
```
maybe create a mixin for all controllers that need a count? or consider changing it to all belongs_to attribute a host has?

lzap commented on app/controllers/concerns/find_common.rb:
```diff
@@ -54,4 +54,8 @@ def resource_class_for(resource)
+
+  def find_hosts_count
+    @hosts_count = Host::Managed.reorder('').authorized.group("#{resource_name}_id").count
+  end
```
Thanks for bringing this up. My first observation is explicit vs implicit. Do we really need to fetch the count for each individual index page? Are you sure all index pages use it? I'd rather prefer a helper method `resource_count` that would cache the number for the lifetime of request. Meaning fetching it only when it's really needed.

tbrisker replied:
This is only run on those index pages that display a host count.

lzap replied:
I understand but it leaves space for errors in the future. Explicit "pull" mechanism might be safer.

lzap commented on lib/tasks/fix_cached_counters.rake:
```diff
@@ -1,24 +1,11 @@
   puts "LookupKey corrected"
-  Hostgroup.all.each{|el| Hostgroup.reset_counters(el.id, :hosts)}
-  puts "Hostgroup corrected"
 end
```
Are you leaving some counters on purpose?

tbrisker replied:
yes, the puppet class related counters don't face the same issue with deadlocks as the host ones, but we can consider getting rid of them as well. Not sure it has to be in the same pr.

lzap replied:
> yes, the puppet class related counters don't face the same issue with deadlocks as the host ones, but we can consider getting rid of them as well. Not sure it has to be in the same pr.

I think for consistency, we should do it in one PR. Just rename issue/PR
and go for it. The theme is simply dropping all cache counters from
here.

I will also do testing in beaker against 100+ real hosts for this patch.

## 

Later,
 Lukas #lzap Zapletal

beav commented:
@tbrisker many thanks for this patch :koala:

I'm not super familiar with the foreman codebase but it looks good to me. I will try to apply this to a instance with a larger DB and test.

I'm planning on checking the following areas:
- host registration
- architectures page
- operating systems page
- host related pages that have counts

Is there anything else I should hit?

tbrisker commented:
@beav I'd suggest comparing load times before/after for all of the index pages touched by this PR, as some use different methods of calculating the counts (for example, config_groups calculate host count recursively via hostgroups).

beav commented:
@tbrisker ah i see what you mean, will do

beav commented:
@tbrisker I did not get a chance to try this out today with a larger db. I will give it a shot on Monday.

beav commented:
I applied this to a server with 2K hosts that were already registered and did not notice any slowdowns. For example:

```
2016-06-20 12:47:21 EDT LOG:  duration: 1.612 ms  execute <unnamed>: SELECT COUNT(*) AS count_all, architecture_id AS architecture_id FROM "hosts"  WHERE "hosts"."type" IN ('Host::Managed') AND "hosts"."organization_id" IN (1) GROUP BY architecture_id
```

I tested this on a shared machine so I did not exercise host groups, config groups or realms but the other pages looked fine.

ohadlevy commented:
IMHO 2K or 5K sample size is irrelevant, this usually manifests itself with
ACTIVE db (e.g. many puppet runs / facts imports / reports /enc queries).
I would try on a db with 10K or more, ideally around 50K hosts.

On Mon, Jun 20, 2016 at 7:54 PM, Chris Duryee notifications@github.com
wrote:

> I applied this to a host with 2K hosts and did not notice any slowdowns.
> For example:
> 
> 2016-06-20 12:47:21 EDT LOG:  duration: 1.612 ms  execute <unnamed>: SELECT COUNT(*) AS count_all, architecture_id AS architecture_id FROM "hosts"  WHERE "hosts"."type" IN ('Host::Managed') AND "hosts"."organization_id" IN (1) GROUP BY architecture_id
> 
> I tested this on a shared machine so I did not exercise host groups,
> config groups or realms but the other pages looked fine.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/theforeman/foreman/pull/3596#issuecomment-227201754,
> or mute the thread
> https://github.com/notifications/unsubscribe/AABOx3dgDEpb1QNVH_18c77K4qiNp3kSks5qNsXcgaJpZM4I3VTo
> .

lzap commented:
If we hit any performance issues, we should definitely not try to solve them with this evil idea of counters but with Rails cache and fragment cache. For example dashboard is ideal use case for caching - it should only update once per 10 minutes by default, or something like that.

My hosts counts I get with my Satellite 6.2 installation is irrelevant - I have about 100 systems registered. So I won't even bother testing this.

ohadlevy commented:
if you get all of your 100 hosts concurrently hitting the API's you might
reproduce it, the large number of hosts only imply the concurrent clients

beav commented:
@lzap i agree, I would rather address any slowness that occurs via other means than counter cache.

In addition to the ways you mention, we may also be able to increase postgres shared buffers to reduce IO on count calls. There are also some tips at https://wiki.postgresql.org/wiki/Slow_Counting and https://wiki.postgresql.org/wiki/Index-only_scans.

@ohadlevy  Is there an example foreman db with 50K hosts available? I can try loading it up and testing some of the new count queries.

From a functionality testing perspective, I did not find any issue with this patch when I used it.

tbrisker commented:
Tested this with a db of ~40k hosts. All index pages load within 1-3 seconds when running both the server and the db on my measly laptop in dev environment. I assume any production environment managing such a large infrastructure will run on much more powerful hardware which will push these numbers even lower.
In other words, looks like the counter cache was a premature optimization and can be safely removed as long as we're ok with losing the search&sort functionality based on the counts (or making it more difficult to implement).

tbrisker commented:
Pushed a new version that fixed a couple of permissions issues. Still runs in <3sec for ~40k hosts, even when viewing with a user that has a host filter in place (and no longer knows how many hosts there are that they aren't supposed to see)

tbrisker commented:
Updated to use a lazy helper to load the counts, changed taxonomies to use same method and added minor optimization in the config group search. I will create a separate commit to remove the remaining counters.

tbrisker committed:
Refs #12193 - Remove remaining cached counters

tbrisker commented:
@lzap ping?

lzap commented:
Will take a look this week, sorry for the delay.

lzap commented on app/models/puppetclass.rb:
```diff
@@ -110,8 +107,8 @@ def all_host_ids
-  def update_total_hosts
-    update_attribute(:total_hosts, all_host_ids.size)
+  def hosts_count
+    all_host_ids.count
```
This is full load into memory (pluck, uniq and others), replace with ordinary SQL call here.

lzap replied:
Hmm, the code that builds `all_host_ids` is quite complex, but it might be possible to optimize it a bit - at least do two queries and sum them up.

tbrisker replied:
It's going to be more then 2 queries because of the hostgroup ancestry (a host can get a puppetclass from it's hostgroup's parent's parent's config group for example), but I changed this to push what I can into the database.

lzap commented on app/services/classification/base.rb:
```diff
@@ -27,10 +27,8 @@ def class_parameters
-      end.map(&:path_elements).flatten(1).uniq
+      # the inner join makes sure we take only keys with actual values
+      class_parameters.joins(:lookup_values).flat_map(&:path_elements).uniq
     end
```
I can find a unit test for this one, is there coverage of this change?

tbrisker replied:
This is a helper function used for calculating which matchers apply to a certain host, it is exercised indirectly by the classification test.

lzap commented on app/views/config_groups/index.html.erb:
```diff
@@ -8,17 +8,17 @@
+      <th><%= _('Puppet classes') %></th>
+      <th><%= _('Hosts') %></th>
+      <th><%= _('Host groups') %></th>
       <th><%= _('Actions') %></th>
```
We lose sorting on these columns (and for other resources as well) but I agree this is the correct approach.

lzap commented on app/views/puppetclasses/index.html.erb:
```diff
@@ -25,9 +25,9 @@
-        <td> <%= link_to puppetclass.total_hosts, hosts_path(:search => "class = #{puppetclass.name}")%></td>
-        <td><%= puppetclass.global_class_params_count %> </td>
-        <td><%= puppetclass.variable_lookup_keys_count %> </td>
+        <td> <%= link_to puppetclass.hosts_count, hosts_path(:search => "class = #{puppetclass.name}")%></td>
```
If the space character is intentional, make it `&nbsp;` please.

tbrisker replied:
fixed.

lzap commented on app/views/variable_lookup_keys/index.html.erb:
```diff
@@ -14,7 +14,7 @@
         <td class="display-two-pane ellipsis"><%= link_to_if_authorized lookup_key.key, hash_for_edit_variable_lookup_key_path(:id => lookup_key).merge(:auth_object => lookup_key, :permission => 'edit_external_variables', :authorizer => authorizer) %></td>
         <td class="ellipsis"><%= link_to_if_authorized lookup_key.puppetclass, hash_for_edit_puppetclass_path(:id => lookup_key.puppetclass).merge(:auth_object => lookup_key.puppetclass, :authorizer => @puppetclass_authorizer) %></td>
-        <td><%=h lookup_key.lookup_values_count %></td>
+        <td><%= lookup_key.lookup_values.count %></td>
```
Just to confirm the "h" function was removed intentionally.

tbrisker replied:
yes, this is just a number, no point in html escaping it.

lzap commented on db/migrate/20140828111505_fix_counters.rb:
```diff
@@ -1,7 +1,7 @@
   # Fix all the cached counters that were corrupted by #5692 and related bugs
   def up
-    Rake::Task['fix_cached_counters'].invoke
+    #Rake::Task['fix_cached_counters'].invoke
```
Please go ahead and remove all commented lines completely from all migrations.

lzap commented on db/migrate/20160616074718_remove_host_counter_cache.rb:
```diff
@@ -0,0 +1,28 @@
+    add_column :hostgroups, :hosts_count, :integer
+    add_column :models, :hosts_count, :integer
+  end
+end
```
This is great candidate for the new style of migration introduced in Rails 4+ (via `change` method).

lzap commented on db/migrate/20160626085636_remove_puppet_counters.rb:
```diff
@@ -0,0 +1,14 @@
+    add_column :lookup_keys, :lookup_values_count, :integer
+    add_column :config_groups, :config_group_classes_count, :integer
+  end
+end
```
The same as above.

lzap commented:
Ok minor comments, except the missing test coverage of the complex count call. Great work!

tbrisker commented:
@lzap Thanks, I believed I fixed all your comments, please re-review

lzap commented on app/models/puppetclass.rb:
```diff
@@ -87,31 +84,25 @@ def klass
+                   .reorder('')
+                   .eager_load(:hostgroup_classes, :config_groups => [:config_group_classes])
+                   .where("#{self.id} IN (hostgroup_classes.puppetclass_id, config_group_classes.puppetclass_id)")
+                   .uniq
```
Good, few points.

We are breaking the contract, it used to be sorted, now it's not. The method is used once (in a RABL), can you add an optional parameter `unsorted` and use it appropriately to save the extra sorting when doing count?

Have we lost `host_type` from above? It's not part of the new query anymore.

tbrisker replied:
added the parameter, `all_hostgroups` is also used in the ui. 
`host_type` is now automatically applied by the `eager_load`, which rails now knows how to do as we start the query from the hostgroup rather then the config group as it was before.

lzap commented on app/models/puppetclass.rb:
```diff
@@ -87,31 +84,25 @@ def klass
+                       .pluck(:id)
+    ids |= Host::Managed.authorized.where(:hostgroup_id => all_hostgroups.map(&:id)).pluck(:id)
+    ids.size
   end
```
Can you do SQL count on line 98 and then simply provide `.pluck.size` alternative when it's zero?

Under what circumstance the line 98 returns nil/false? I can't see it. To me it looks this is always an array (e.g. empty one).

tbrisker replied:
`|=` is set union for arrays - I think you confused it with `||=`. This is because we might get duplicate hosts in the two queries - which is why we can't count in them either. Anyways, I'll change these two queries to run in one, with a bit of extra sql.

lzap replied:
Oh yeah, misread it. Looks good, my last point: I don't see `Puppetclass#hosts_count` covered in tests, can you either show me a case or add few tests?

lzap commented:
Ok, this is it. Few minor questions above, test failure is unrelated. Please rebase onto `develop` when possible and I will start with final test and merge once we are green.

tbrisker commented:
@lzap fixed your comments, test failure unrelated and not quite sure what code climate wants.

lzap commented:
Codeclimate is just a nonsence, I will ignore it (Similar code found in 2 other locations).

tbrisker commented:
@lzap added tests for puppet class host count, thanks

lzap commented:
Merged as fef6cff7e2c3d87c1aac8d0a5fbfd6cae1447a84, thank you!
